libs/api/database/src/lib/content/stores/tags.store.ts

```diff
@@ -0,0 +1,167 @@
+    async findOne(id: string): Promise<TagsDocument> {
+        return this.tags.findById(id);
+    }
+
+    /**
+     * Returns the given tag and all its descendants.
+     * @param king
+     * @param tagId
+     */
+    async fetchDescendants(tagId: string): Promise<TagsTree> {
+        const parentTag = await this.tags.findById(tagId);
+        if (!parentTag) {
+            throw new NotFoundException("No tag with that ID found.");
+        }
+
+        const root: TagsTree = {
+            ...parentTag.toObject(),
+            children: []
+        };
+        const tree = await this.buildDescendantTree(root);
+
+        return tree;
+    }
+
+    private async buildDescendantTree(tagNode: TagsTree): Promise<TagsTree> {
+        const currentRoot = await this.populateImmediateChildren(tagNode._id);
+        if (!currentRoot.children || currentRoot.children.length === 0) {
+            // If we can't find any children for this node, it is a leaf. Return it.
+            return tagNode;
+        }
+
+        // Otherwise, iterate through each child node, find _its_ children,
+        // and add them to the node's `children` property.
+        for (const child of currentRoot.children) {
+            child.children = (await this.buildDescendantTree(child))?.children;
+        }
+
+        return currentRoot;
+    }
+
+    /**
+     * Returns the Tag, with an extra `children` property array that contains all of this
+     * tag's immediate children.
+     * @param id The ID of the tag to populate.
+     */
+    private async populateImmediateChildren(id: string): Promise<TagsTree> {
+        // We force maxDepth to be 0 so that we only get the node's immediate children.
+        // If we went further, Mongo would just give us a flat array, and we'd have to
+        // put it together ourselves anyway.
+        const results: TagsTree[] = await this.tags.aggregate<TagsTree>([
+            { $match: { _id: id } }
+        ])
+        .graphLookup({
+            from: 'tags',
+            startWith: '$_id',
+            connectFromField: '_id',
+            connectToField: 'parent',
+            as: 'children', // Careful: this string should match the name of the property in `TagsTree`.
+            maxDepth: 0,
+        }).exec();
+        if (results.length === 0) {
+            return null;
+        }
+
+        // Because we specify maxDepth of 0, and look up a single ID, we don't
+        // expect (or care about) anything beyond the first result.
+        return results[0];
+    }
+
+    /**
+     * Creates a new tag.
+     * @param form
+     * @param tagKind
+     */
+    async createTag(form: TagsForm, tagKind: TagKind): Promise<TagsDocument> {
+        if (form.parent) {
+            // Ensure the parent exists and its kind matches
+            const parentTag = await this.tags.findById(form.parent);
+            if (!parentTag) {
+                throw new NotFoundException("Could not find any parent tag with that ID.");
+            }
```

Comment:
I tried to guard against edge cases here, but I feel really suspicious I'm not thinking of some.

Reply:
Nothing I can think of.

Reply:
i can't think of anything either